library/core/src/array/mod.rs

```diff
@@ -947,28 +904,73 @@ impl<T, const N: usize> Guard<'_, T, N> {
     }
 }
 
-impl<T, const N: usize> Drop for Guard<'_, T, N> {
+impl<T> Drop for Guard<'_, T> {
     fn drop(&mut self) {
-        debug_assert!(self.initialized <= N);
+        debug_assert!(self.initialized <= self.array_mut.len());
 
         // SAFETY: this slice will contain only initialized objects.
         unsafe {
             crate::ptr::drop_in_place(MaybeUninit::slice_assume_init_mut(
-                &mut self.array_mut.get_unchecked_mut(..self.initialized),
+                self.array_mut.get_unchecked_mut(..self.initialized),
             ));
         }
     }
 }
 
-/// Returns the next chunk of `N` items from the iterator or errors with an
-/// iterator over the remainder. Used for `Iterator::next_chunk`.
+/// Pulls `N` items from `iter` and returns them as an array. If the iterator
+/// yields fewer than `N` items, `Err` is returned containing an iterator over
+/// the already yielded items.
+///
+/// Since the iterator is passed as a mutable reference and this function calls
+/// `next` at most `N` times, the iterator can still be used afterwards to
+/// retrieve the remaining items.
+///
+/// If `iter.next()` panicks, all items already yielded by the iterator are
+/// dropped.
+///
+/// Used for [`Iterator::next_chunk`].
 #[inline]
-pub(crate) fn iter_next_chunk<I, const N: usize>(
-    iter: &mut I,
-) -> Result<[I::Item; N], IntoIter<I::Item, N>>
-where
-    I: Iterator,
-{
-    let mut map = iter.map(NeverShortCircuit);
-    try_collect_into_array(&mut map).map(|NeverShortCircuit(arr)| arr)
+pub(crate) fn iter_next_chunk<T, const N: usize>(
```

Comment:
I am curious about what motivated your reasoning to change the type signature here?

Reply:
@emilHof It's essentially the same; I didn't *have* to change it.

This is just a bit more consistent with how I did the other things in this file.  (Like how `try_from_fn_erased` takes `impl FnMut`, not `F` … `where F: FnMut`.)  I tend to think of APIT like this as "generic elision" -- for things like iterators and closures, I tend not to actually care about the specific type, so having a used-once generic for it where I have to read the bounds to follow the signature is kinda meh.

I like thinking of it as "this function takes a mutable reference to an iterator of `T`s, and returns an array of `T`s".  Like how you might write a function taking `Iterator<T>` in Java, or an `IEnumerable<T>` in C#, just without the `dyn` indirection that that results in those languages.  Thinking of it as "This takes an `I`, and returns an array of the item type of `I`, and BTW `I` needs to be an iterator" is just less direct in my brain.  The item type is the important one, not the iterator's type, for me.

TL/DR: Entirely preference.

Reply:
@scottmcm That is a really explanation!

Your reasoning definitely makes sense. In a way, the `T` is maybe more relevant to the purpose of these methods, so having the `T` and not some "elusive" `I` in the signature definitely seems very sensible! At least, this is how I am interpreting what you are saying.

Thank you for the great response! :)

Reply:
Yup, that's a good phrasing of it!